Multimethods

malcolmsparks opened the pull request:
Tested against the JUXT website.

Commit:
Replace read-krei-files with individual read-krei-filen

This trival change is to support holding on to the original krei file for passing to
build components, so debugging information has access to the krei-file
resource URL.

Commit:
Factor out Figwheel and Sass via multimethods

This is a backward compatible change that retains the existing
functionality while preparing the way for new 'providers' of build
functionality, such as shadow-cljs and JIT image resizing.

malcolmsparks commented:
I forgot to mention that the result of the `init!` call will get passed to the `notify!` calls. The only rationale so far for doing this is that you were hardcoding `./target` in the Sass builder. I thought I'd fix that for you.

Commit:
Avoid re-parsing krei-files in `build-sass`

This was an oversight.

Commit:
Pass build config to watch

This is a new design whereby the build configuration is passed to
krei/kick, for example, from an Aero configuration.

```clojure
  {:default {}
   :dev
   {:kick/builder
    {:kick.builder/target "target"

     :kick/sass {:sources ["public/app.scss"]}

     :kick/figwheel {:builds [{:id "main"
                               :figwheel {:on-jsload "edge.main/init"}
                               :compiler {:main edge.main
                                          :output-to "public/edge.js"
                                          :output-dir "public/edge.out"
                                          :asset-path "/public/edge.out"}}]}}}}
```

The watch and returning close function correspond directly to
`integrant.core/init-key` and `integrant.core/halt-key!`, although
there is no integrant dependency from krei/kick

IMPORTANT: Since krei can _no longer build independently_ the
production build (called by main) has been removed entirely and
consuming projects now need to provide this functionality.

malcolmsparks commented:
OK, this is now ready to review. There is more context and design rationale in the individual commit messages. I've gone for a BYOD (bring your own dependency) policy for the individual providers, and none are therefore privileged. I'm planning to add shadow-cljs in due course. The point of this BYOD policy is to avoid dependency bloat (avoid paying for the cost of dependencies you don't use).

One strange thing is that Edge seems to 'work' (build the sass) without seeming to bring in 

```clojure
deraen/sass4clj {:mvn/version "0.3.1"}
```
(not sure what's going on there, but it doesn't affect this review).

SevereOverfl0w commented on src/io/dominic/krei/alpha/core.clj:
```diff
@@ -1,18 +1,41 @@
-    [io.dominic.krei.alpha.impl.debounce :as krei.debounce]))
+   [clojure.java.classpath :as classpath]
+   [clojure.string :as string]
+   [clojure.tools.logging :as log]
```
tools.logging isn't a dependency of the project. So this commit only compiles by coincidence.

SevereOverfl0w commented on src/io/dominic/krei/alpha/core.clj:
```diff
@@ -96,36 +117,15 @@
-                                 (map #(update-in % [:compiler :output-to] (comp str target-relative))))
-                           krei-files)}))
-    (build-sass)
+    ;; @dmc what's this? still something we need?
```
which this? :p Presuming you mean ` (run! dirwatch/close-watcher krei-builders)` it takes care of shutting down the file watchers.

SevereOverfl0w reviewed (changes requested):
A few minor things. Overall looks good.

I don't see a way to perform a one-time build of clojurescript with this system, that would be a big gap.

Commit:
Disable shadow-cljs due to conflict with Figwheel

```
----  Exception    ----
  No implementation of method: :-find-sources of protocol: #'cljs.closure/Compilable found for class: cljs.build.api$inputs$reify__6688
----  Exception Stack Trace  ----
java.lang.IllegalArgumentException: No implementation of method: :-find-sources of protocol: #'cljs.closure/Compilable found for class: cljs.build.api$inputs$reify__6688
 at clojure.core$_cache_protocol_fn.invokeStatic (core_deftype.clj:583)
    clojure.core$_cache_protocol_fn.invoke (core_deftype.clj:575)
    cljs.closure$eval45293$fn__45307$G__45284__45314.invoke (cljs_closure.clj:475)
    cljs.closure$build.invokeStatic (cljs_closure.clj:2707)
    cljs.closure$build.invoke (cljs_closure.clj:2672)
    cljs.build.api$build.invokeStatic (api.clj:208)
    cljs.build.api$build.invoke (api.clj:189)
    figwheel_sidecar.components.cljs_autobuild$cljs_build.invokeStatic (cljs_autobuild.clj:28)
```

Commit:
Update README and LICENSE

Commit:
Fix typo

Commit:
Move package structure to align with renaming

Commit:
Rename alias and delete redundant comment

SevereOverfl0w commented on src/io/dominic/krei/alpha/providers/shadow_cljs.clj.txt:
```diff
@@ -0,0 +1,36 @@
+;; Suggested dependencies:
+;; thheller/shadow-cljs {:mvn/version "2.0.123"}
+;; As there's a bug in tools.deps.alpha (TDEPS-26), also add:
```
TDEPS-26 is fixed https://dev.clojure.org/jira/browse/TDEPS-26

Commit:
Fix debounce to ignore Emacs temp files

Commit:
Reload on adoc file change (in addition to html file change)

This calls for configurability.

Commit:
Fix docstring

Commit:
Build sass on init

This is to ensure that unpacked production instances will build their own
CSS prior to starting.

Commit:
Respect user setting of cljs optimizations

Commit:
Prevent error from failing agent

Commit:
Parameterize server-port

Commit:
Add pom.xml

Commit:
Update to multiple Sass builds

Syntax has changed - see Edge

Commit:
Warn if Sass file not found

SevereOverfl0w committed:
Add figwheel alias

This makes developing with figwheel possible without loading kick in the
context of a project.

SevereOverfl0w committed:
Update README to use one sentence per-line

This is a convention set by Asciidoctor.

SevereOverfl0w committed:
Remove invalid cljs.build.api reference

SevereOverfl0w committed:
Add build-once support

Commit:
Reformatting and logging

Commit:
Replace `:sources` with `:source` and `:target`

This BINARY INCOMPATIBLE change is to allow finer control of the
target filename. This is required by Edge to place generated CSS files
into a directory named 'public'.

Given the fast-moving alpha status of Kick, with the ease of fixing its
(few) current dependants, simplicity has been chosen over binary
compatibility.

Commit:
Merge new oneshot! code with new build-sass semantic

Commit:
Delete empty file, causing tools.namespace to NPE!

SevereOverfl0w committed:
Add namespace loading guard

This prevents a bug which happens as a combination of:
- Kick being a git dependency
- Kick allowing users to bring their own dependencies
- tools.namespace loading all namespaces which are in directories

This guard prevents the namespace from exploding when required without
the required dependency present.  This is not completely ideal, due to
the confusion it will push down onto users when namespaces unexpectedly
don't work.

SevereOverfl0w committed:
Add support for figwheel-main

SevereOverfl0w committed:
Fix notify for sass

SevereOverfl0w committed:
Make figwheel main's halt! idempotent

SevereOverfl0w committed:
Fix target directory comparison

SevereOverfl0w committed:
Remove pack from watch-dirs for figwheel main

This causes a compilation error when starting a clojurescript REPL, as
figwheel-main attempts to compile the code in watch-dirs then.

SevereOverfl0w committed:
Correct function name

SevereOverfl0w committed:
Remove debugging pprint

SevereOverfl0w committed:
Fix lib directory filtering

SevereOverfl0w committed:
Fix kick debounce agent failure condition

SevereOverfl0w committed:
Add target relativization to cljs :source-map

SevereOverfl0w committed:
Disable cljs temp dir when using source map

You need the :output-dir to be produced when source maps are enabled.

SevereOverfl0w committed:
Update to clojure.basis

clj no longer produces a clojure.libfile, and kick needs this to remove
itself from the classpath when using figwheel-main.  Newer versions of
clj instead provide a clojure.basis, which is vaguely the same as the
old basis.

Supports both in order to handle the transition period of updates.

Commit:
Merge pull request #1 from SevereOverfl0w/clojure.basis

Update to clojure.basis
